provider/azure/credentials_test.go

```diff
@@ -62,9 +73,243 @@ func (s *credentialsSuite) TestServicePrincipalSecretHiddenAttributes(c *gc.C) {
 	envtesting.AssertProviderCredentialsAttributesHidden(c, s.provider, "service-principal-secret", "application-password")
 }
 
-func (s *credentialsSuite) TestDetectCredentials(c *gc.C) {
+func (s *credentialsSuite) TestDetectCredentialsNoAccounts(c *gc.C) {
 	_, err := s.provider.DetectCredentials()
 	c.Assert(err, jc.Satisfies, errors.IsNotFound)
+	calls := s.azurecliExecutor.Calls()
+	c.Assert(calls, gc.HasLen, 1)
+	c.Assert(calls[0].FuncName, gc.Equals, "ListAccounts")
+}
+
+func (s *credentialsSuite) TestDetectCredentialsListError(c *gc.C) {
+	s.azurecliExecutor.SetErrors(errors.New("test error"))
+	_, err := s.provider.DetectCredentials()
+	c.Assert(err, jc.Satisfies, errors.IsNotFound)
+}
+
+func (s *credentialsSuite) TestDetectCredentialsOneAccount(c *gc.C) {
+	s.azurecliExecutor.Accounts = []azurecli.Account{{
+		CloudName: "AzureCloud",
+		ID:        "test-account-id",
+		IsDefault: true,
+		Name:      "test-account",
+		State:     "Enabled",
+		TenantId:  "tenant-id",
+	}}
+	s.azurecliExecutor.Clouds = []azurecli.Cloud{{
+		Endpoints: azurecli.CloudEndpoints{
+			ActiveDirectoryGraphResourceID: "https://graph.invalid",
+			ResourceManager:                "https://arm.invalid",
+		},
+		IsActive: true,
+		Name:     "AzureCloud",
+	}}
+	cred, err := s.provider.DetectCredentials()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(cred, gc.Not(gc.IsNil))
+	c.Assert(cred.DefaultCredential, gc.Equals, "test-account")
+	c.Assert(cred.DefaultRegion, gc.Equals, "")
+	c.Assert(cred.AuthCredentials, gc.HasLen, 1)
+	c.Assert(cred.AuthCredentials["test-account"].Label, gc.Equals, "AzureCloud subscription test-account")
+
+	calls := s.azurecliExecutor.Calls()
+	c.Assert(calls, gc.HasLen, 4)
+	c.Assert(calls[0].FuncName, gc.Equals, "ListAccounts")
```

Comment:
I wonder if it might look nicer and give more confidence if we mocked at the command-executor level here, for example by providing a map from command+args to output. But that's probably me and my general meh'ness about the mock test stuff.